src/main/java/net/snowflake/client/core/bind/BindUploader.java

```diff
@@ -0,0 +1,438 @@
+
+  private static final String PUT_STMT = "PUT"
+      + " 'file://%s/*'"            // argument 1: folder
+      + " '%s'"                     // argument 2: stage path
+      + " parallel=10"              // upload chunks in parallel
+      + " overwrite=true"           // skip file existence check
+      + " auto_compress=false"      // we compress already
+      + " source_compression=gzip"; //   (with gzip)
+
+  private static final int PUT_RETRY_COUNT = 3;
+
+  // session of the uploader
+  private final SFSession session;
+
+  // fully-qualified stage path to upload binds to
+  private final String stagePath;
+
+  // local directory to write binds to
+  private final Path bindDir;
+
+  // whether the uploader has completed
+  private boolean closed = false;
+
+  // size (bytes) per file in upload, 100MB default
+  private long fileSize = 100 * 1024 * 1024;
+
+  /**
+   * Create a new BindUploader which will write binds to the *existing*
+   * bindDir and upload them to the given stageDir
+   * @param stageDir
+   * @param bindDir
+   */
+  private BindUploader(SFSession session, String stageDir, Path bindDir)
+  {
+    this.session = session;
+    this.stagePath = "@" + STAGE_NAME + "/" + stageDir;
+    this.bindDir = bindDir;
+  }
+
+  /**
+   * Create a new BindUploader which will upload to the given stage name
+   * Ensures temporary directory for file writing exists
+   * @param stageDir
+   * @return BindUploader
+   * @throws BindException
+   */
+  public synchronized static BindUploader newInstance(SFSession session, String stageDir)
+      throws BindException
+  {
+    try
+    {
+      Path bindDir = Files.createTempDirectory(PREFIX);
+      return new BindUploader(session, stageDir, bindDir);
+    }
+    catch(IOException ex)
+    {
+      throw new BindException(
+          String.format("Failed to create temporary directory: %s", ex.getMessage()));
+    }
+  }
+
+  /**
+   * Upload the bindValues to stage
+   * @param bindValues
+   * @throws BindException
+   */
+  public void upload(Map<String, ParameterBindingDTO> bindValues) throws BindException
+  {
+    if (!closed)
+    {
+      serializeBinds(bindValues);
+      putBinds();
+    }
+  }
+
+  /**
+   * Save the binds to disk
+   * @param bindValues
+   * @throws BindException
+   */
+  private void serializeBinds(Map<String, ParameterBindingDTO> bindValues) throws BindException
+  {
+    ArrayList<ArrayList<String>> columns = getColumnValues(bindValues);
+    List<String[]> rows = buildRows(columns);
+    writeRowsToCSV(rows);
+  }
+
+  /**
+   * Convert binding map to a list of values for each column
+   * Perform necessary type casts and invariant checks
+   * @param bindValues binding map
+   * @return list of values for each column
+   * @throws BindException
+   */
+  private ArrayList<ArrayList<String>> getColumnValues(Map<String, ParameterBindingDTO> bindValues) throws BindException
```

Comment:
Why returning ArrayList all the time? Will List<List<String>> work? Basically, `List` is the interface and the rest the code should not need to know about ArrayList

Reply:
I was indecisive on that part, mostly because when we build the rows we perform a ton of random accesses on potentially huge lists, which could be slow if the binds happen to use another list implementation. For our case it really doesn't make a difference, since we only use `ArrayList`.

I considered just using arrays for the whole thing to make it simpler, but I wanted to avoid copying each `List` into a `String[]` (in case we have an enormous list of bind values). Which do you prefer?

Reply:
I think it's fine as long as you use ArrayList when instantiate the object. The rest the the code will not need to change the implementation.